api/nodes/serializers.py

```diff
@@ -586,27 +586,55 @@ def get_unregistered_contributor(self, obj):
 
 class NodeContributorsCreateSerializer(NodeContributorsSerializer):
     """
-    Overrides NodeContributorsSerializer to add target_type and required id field
+    Overrides NodeContributorsSerializer to add email, full_name, and non-required users field.
     """
-    id = ContributorIDField(required=True)
-    target_type = TargetTypeField(target_type='users')
+    id = ContributorIDField(source='_id', required=False, allow_null=True)
+    full_name = ser.CharField(required=False)
+    email = ser.EmailField(required=False)
+
+    users = RelationshipField(
+        related_view='users:user-detail',
+        related_view_kwargs={'user_id': '<pk>'},
+        required=False
+    )
 
     def create(self, validated_data):
-        auth = Auth(self.context['request'].user)
+        id = validated_data.get('_id')
+        email = validated_data.get('email')
         node = self.context['view'].get_node()
-        contributor = get_object_or_error(User, validated_data['_id'], display_name='user')
-        # Node object checks for contributor existence but can still change permissions anyway
-        if contributor in node.contributors:
-            raise exceptions.ValidationError('{} is already a contributor'.format(contributor.fullname))
-
-        bibliographic = validated_data['bibliographic']
+        auth = Auth(self.context['request'].user)
+        full_name = validated_data.get('full_name')
+        bibliographic = validated_data.get('bibliographic')
         permissions = osf_permissions.expand_permissions(validated_data.get('permission')) or osf_permissions.DEFAULT_CONTRIBUTOR_PERMISSIONS
-        node.add_contributor(contributor=contributor, auth=auth, visible=bibliographic, permissions=permissions, save=True)
+
+        if not id and not full_name:
```

Comment:
If you want to clean this up a bit, I'd lean towards:

1. Make a validation function that ensures that you have some combination of the correct information and, if not, raises the appropriate error.
2. Make a model method that will create a contributor regardless of which way it gets made and have that do the creation logic. Something like `add_contributor_registered_or_not(user_id, full_name, email, auth, bibliographic, permissions, save)`. Then most of this function gets reduced to "validate; create;" rather than heavy logic. The validation tests can happen as TestCase unit tests (which are fast) rather than as view tests, and you can do at most one view test to verify that your validation function is being used. 

Your model method will probably be a little easier to write like that is my guess. Maybe not a *lot* easier, but.